pipe_anchorages/transforms/voyages_create.py

```diff
@@ -0,0 +1,62 @@
+import apache_beam as beam
+import datetime as dt
+import pytz
+
+def create_voyage(visits):
+    # first voyage start from 'we do not know' = None
+    visit_before = {2: None, 3: None, 4: None}
+
+    def min_confidence(v, c:int)->bool:
+        return v['confidence'] >= c
+
+    def has_duration(v1, v2):
+        return v1 == None or v1['end_timestamp'] < v2['start_timestamp']
+
+    for visit in visits:
+        for c in [2,3,4]:
+            if (visit and min_confidence(visit, c) and has_duration(visit_before[c],visit)) or (not visit and visit_before[c]):
+                yield build_voyage(c, visit_before[c], visit)
+                visit_before[c] = visit
+
+def build_voyage(trip_confidence, visit_before, visit_after):
+    voyage_ssvid = visit_before['ssvid'] if visit_before else visit_after['ssvid']
+    voyage_vessel_id = visit_before['vessel_id'] if visit_before else visit_after['vessel_id']
+    voyage_trip_start = visit_before['end_timestamp'] if visit_before else None
+    voyage = {
+        'trip_confidence': trip_confidence,
+        'ssvid': voyage_ssvid,
+        'vessel_id': voyage_vessel_id,
+        'trip_start': voyage_trip_start,
+        'trip_end': visit_after['start_timestamp'] if visit_after else None,
+        'trip_start_anchorage_id': visit_before['end_anchorage_id'] if visit_before else None,
+        'trip_end_anchorage_id': visit_after['start_anchorage_id'] if visit_after else None,
+        'trip_start_visit_id': visit_before['visit_id'] if visit_before else None,
+        'trip_end_visit_id': visit_after['visit_id'] if visit_after else None,
+        'trip_start_confidence': visit_before['confidence'] if visit_before else None,
+        'trip_end_confidence': visit_after['confidence'] if visit_after else None,
```

Comment:
I think that there are some abstractions here that can be made. You are doing a bunch of `if dict a is not None, then extract some field out of dict a, else use None`. You could create a function for that like this:

```python
def maybe_get(key, dict, default=None):
    return dict[key] if dict else default
```

In some cases you are doing something like `take key k from either a or b, whichever is not None`, which can be something like this:

```python
def get_from_either(key, first_dict, second_dict):
    return first_dict[key] if first_dict else second_dict[key]
```


and then all of these clauses can be instead something like this:

```python
voyage = {
    'ssvid': get_from_either('ssvid', visit_before, visit_after]),
    'vessel_id': get_from_either('vessel_id', visit_before, visit_after),
    'trip_start': maybe_get('end_timestamp', visit_before),
    # ...
}
```